src/use-case/delivery-time/delivery-time.service.ts

```diff
@@ -0,0 +1,153 @@
+    input: CalculateDeliveryTimeInputDto,
+  ): Promise<DeliveryTimeResultDto[]> {
+    const { vechileDetails, packageList } = input
+    const { noOfVechiles, maxSpeed, maxCarriableWeight } = vechileDetails
+
+    // Initialize remaining package list,
+    // at begining is the package list itself
+    let remainingPackageList = [...packageList]
+
+    // Get the delivery combo for each trip
+    // package closest to maxCarriableWeight go first
+    // delivery sequence store in an array
+    const deliveryComboSequence: PackageDtoForTime[][] = []
+    while (remainingPackageList.length > 0) {
+      const combo = getPackageComboPerTrip(
+        remainingPackageList,
+        maxCarriableWeight,
+      )
+
+      deliveryComboSequence.push(combo)
+
+      remainingPackageList = remainingPackageList.filter(
+        (eachPackage) => !combo.find((eachCombo) => eachCombo === eachPackage),
+      )
+    }
+
+    // Initialize an array to represent avaialble vechiles available time
+    let currentVechilesAvailabiltyState: VechileStateDto[] = Array(noOfVechiles)
+      .fill({})
+      .map((_, index) => {
+        return {
+          returningTimeForThisVechile: 0,
+          vechileNo: '0' + (index + 1).toString(),
+        }
+      })
+
+    const packagesEstimatedDeliverTime: PackageDeliveredTimeDto[] = []
+
+    for (const packageComboForThisTrip of deliveryComboSequence) {
+      // Get each package need how many time to deliver
+      const packageNeedTimeList: PackageTimeNeededDto[] =
+        packageComboForThisTrip.map((packageDto) => {
+          const timeNeeded = getShipmentArrivalTime(
+            packageDto.distanceInKm,
+            maxSpeed,
+          )
+
+          return {
+            ...packageDto,
+            timeNeeded,
+          }
+        })
+
+      // assign vechile for this package combo
+      // return the vechile number
+      // the time needed for this vechile to return
+      const {
+        assignedVechileNo,
+        nextAvailableTimeForVechile,
+        currentWaitingTime,
+      } = assignVechileForThisTrip(
+        currentVechilesAvailabiltyState,
+        packageNeedTimeList,
+      )
+
+      // get estimated delivery time of each package
+      // based on the vechile assigned
+      // store in an array
+      packageNeedTimeList.map((packageTimeNeed) => {
+        const estimatedDeliveryTime =
+          currentWaitingTime + packageTimeNeed.timeNeeded
+
+        packagesEstimatedDeliverTime.push({
+          ...packageTimeNeed,
+          estimatedDeliveryTime,
+        })
+      })
+
+      // update the vechile state
+      // cause 1 vechile is assigned to this trip
+      // store return time for this vechile in the state
+      let objIndex = currentVechilesAvailabiltyState.findIndex(
+        (obj) => obj.vechileNo == assignedVechileNo,
+      )
+      currentVechilesAvailabiltyState[objIndex].returningTimeForThisVechile =
+        nextAvailableTimeForVechile
+    }
```

Comment:
restructure this chunk of code [here](https://github.com/kenchoong/courrier-service/pull/9), for better readability and unit test